libsolidity/analysis/ControlFlowGraph.cpp

```diff
@@ -0,0 +1,552 @@
+			m_parser(_parser), m_origBreakJump(_parser.m_breakJump), m_origContinueJump(_parser.m_continueJump)
+		{
+			m_parser.m_breakJump = _breakJump;
+			m_parser.m_continueJump = _continueJump;
+		}
+		~BreakContinueScope()
+		{
+			m_parser.m_breakJump = m_origBreakJump;
+			m_parser.m_continueJump = m_origContinueJump;
+		}
+	private:
+		ControlFlowParser& m_parser;
+		CFGNode* m_origBreakJump;
+		CFGNode* m_origContinueJump;
+	};
+};
+
+}
+}
+
+bool ControlFlowParser::visit(BinaryOperation const& _operation)
+{
+	solAssert(!!m_currentNode, "");
+
+	switch(_operation.getOperator())
+	{
+		case Token::Or:
+		case Token::And:
+		{
+			appendControlFlow(_operation.leftExpression());
+
+			auto nodes = splitFlow<2>();
+			nodes[0] = createFlow(nodes[0], _operation.rightExpression());
+			mergeFlow(nodes, nodes[1]);
+
+			return false;
+		}
+		default:
+			break;
+	}
+	return ASTConstVisitor::visit(_operation);
+}
+
+bool ControlFlowParser::visit(Conditional const& _conditional)
+{
+	solAssert(!!m_currentNode, "");
+
+	_conditional.condition().accept(*this);
+
+	auto nodes = splitFlow<2>();
+
+	nodes[0] = createFlow(nodes[0], _conditional.trueExpression());
+	nodes[1] = createFlow(nodes[1], _conditional.falseExpression());
+
+	mergeFlow(nodes);
+
+	return false;
+}
+
+bool ControlFlowParser::visit(IfStatement const& _ifStatement)
+{
+	solAssert(!!m_currentNode, "");
+
+	_ifStatement.condition().accept(*this);
+
+	if (_ifStatement.falseStatement())
+	{
+		auto nodes = splitFlow<2>();
+		nodes[0] = createFlow(nodes[0], _ifStatement.trueStatement());
+		nodes[1] = createFlow(nodes[1], *_ifStatement.falseStatement());
+		mergeFlow(nodes);
+	}
+	else
+	{
+		auto nodes = splitFlow<2>();
+		nodes[0] = createFlow(nodes[0], _ifStatement.trueStatement());
+		mergeFlow(nodes, nodes[1]);
+	}
+
+	return false;
+}
+
+bool ControlFlowParser::visit(ForStatement const& _forStatement)
+{
+	solAssert(!!m_currentNode, "");
+
+	if (auto initializationExpression = _forStatement.initializationExpression())
+		initializationExpression->accept(*this);
+
+	auto condition = createLabelHere();
+
+	if (auto conditionExpression = _forStatement.condition())
+		appendControlFlow(*conditionExpression);
+
+	auto loopExpression = newLabel();
+	auto nodes = splitFlow<2>();
+	auto& afterFor = nodes[1];
+	m_currentNode = nodes[0];
+
+	{
+		BreakContinueScope scope(*this, afterFor, loopExpression);
+		appendControlFlow(_forStatement.body());
+	}
+
+	placeAndConnectLabel(loopExpression);
+
+	if (auto expression = _forStatement.loopExpression())
+		appendControlFlow(*expression);
+
+	connect(m_currentNode, condition);
+	m_currentNode = afterFor;
+
+	return false;
+}
+
+bool ControlFlowParser::visit(WhileStatement const& _whileStatement)
+{
+	solAssert(!!m_currentNode, "");
+
+	if (_whileStatement.isDoWhile())
+	{
+		auto afterWhile = newLabel();
+		auto whileBody = createLabelHere();
+
+		{
+			BreakContinueScope scope(*this, afterWhile, whileBody);
```

Comment:
Actually - does `continue` really go to `whileBody` and not to `condition`? Might that be a bug in the code generator?

Reply:
This is a ``do { ... } while(...);``. For those ``continue`` goes back to the body and that's what I would expect, so I don't think it's a bug. In the ``while(...) { ... }`` case it does indeed go to ``whileCondition`` instead.

Reply:
Turns out, in C and Javascript the `continue` statement jumps to right before the condition, while in Solidity, it jumps to the beginning of the loop body. We should change that with 0.5.0.